### ISSUE TITLE
Add support for listening and reconfiguring the keyboard settings from locale1

### ISSUE THREAD

**Issue by Conan-Kudo** [canonical/mir#3267]
### Proposal

Please add support for listening and dynamically configuring keyboard settings based on [`org.freedesktop.locale1`](https://www.freedesktop.org/software/systemd/man/latest/org.freedesktop.locale1.html).

### Use-cases

This is needed in order to handle cases like keyboard layout switching from the login manager (e.g. SDDM) or the installer (e.g. Calamares or Anaconda).

### References

Here are a couple of example implementations for other compositors:

* KWin: https://invent.kde.org/plasma/kwin/-/merge_requests/2921
* Sway: https://github.com/alebastr/sway-systemd/blob/main/src/locale1-xkb-config

**Comment by Conan-Kudo** [comment]
This is going to be needed for resolving [rhbz#2278864](https://bugzilla.redhat.com/show_bug.cgi?id=2278864) for Fedora Miracle WM Spin releasing with Fedora 41.

cc: @mattkae

**Comment by Conan-Kudo** [comment]
This is really needed with Fedora 42 now that LXQt is also leveraging Mir through Miriway.

cc: @AlanGriffiths

**Comment by AlanGriffiths** [comment]
I know it doesn’t handle all cases, but launching miriway-shell via a login shell at least picks up the session settings.

This shouldn’t be hard, but there’s limited bandwidth. (PRs welcome)

**Comment by AlanGriffiths** [comment]
I've done a little investigation:

The Mir support needed is already in place with `miral::Keymap::set_keymap()` - a shell (such as Miriway) would need to add the dbus listener

**Comment by AlanGriffiths** [comment]
After a bit of puzzlement I have concluded (testing on 24.04) that locale changes are not actually notified over dbus.

I can read the settings before and after, but don't get a "PropertiesChanged" notification when I change them.

```
$ dbus-send --system --dest=org.freedesktop.DBus --type=method_call --print-reply /org/freedesktop/DBus org.freedesktop.DBus.ListNames | grep -i -e properties -e locale
$
```

**Comment by AlanGriffiths** [comment]
Hmm. I don't have a native install of fedora right now, but in a LXD container I use for testing I have similar problems. Maybe it is me?

**Comment by AlanGriffiths** [comment]
After a bit of a rabbithole I now have the dbus side understood.

But the question arises how to integrate with the existing configuration...

Currently, `miral::Keymap` defaults to keyboard settings pulled from udev:

```c++
    input_enumerator.match_subsystem("input");
    input_enumerator.scan_devices();

    for (auto& device : input_enumerator)
    {
        if (auto const layout  = device.property("XKBLAYOUT"))
        {
            auto const options = device.property("XKBOPTIONS");
            auto const variant = device.property("XKBVARIANT");
```

And allows this to be overridden by the `keymap` setting derived from either "static" or "live" configuration.

Are we expecting [org.freedesktop.locale1](https://www.freedesktop.org/software/systemd/man/latest/org.freedesktop.locale1.html) to override any of this? Or supplement it?

Note that "`SetX11Keyboard()` simply sets a default that may be used by later sessions" which implies that `org.freedesktop.locale1` supplies a default, not something to be subscribed to?

**Comment by Conan-Kudo** [comment]
> Are we expecting [org.freedesktop.locale1](https://www.freedesktop.org/software/systemd/man/latest/org.freedesktop.locale1.html) to override any of this? Or supplement it?
> 
> Note that "`SetX11Keyboard()` simply sets a default that may be used by later sessions" which implies that `org.freedesktop.locale1` supplies a default, not something to be subscribed to?

Remember that the open source world uses the word "default" incorrectly quite often. That said, the desired behavior is up to us.

For GNOME, the idea is that if there is no setting by the desktop, it inherits and responds to updates from `locale1`. KWin has a flag to switch to subscribe to it instead. And so on.

**Comment by AlanGriffiths** [comment]
> Remember that the open source world uses the word "default" incorrectly quite often. That said, the desired behavior is up to us.

Yes. The question is which behavior would satisfy the author of this issue?

I had no recollection of the reason we're checking udev (although git is convincing that I once wrote the code). However, after investigating, it comes from [a discussion](https://github.com/canonical/mir/pull/372#pullrequestreview-120894762)  with @RAOF.

The easiest implementation (that avoids changing existing logic) is for downstream to explicitly ask to track `locale1` instead of the existing behavior and that's what I have in the draft PR. Is that good for everyone? (I can implement "a switch" in Miriway)

**Comment by Conan-Kudo** [comment]
Fine with me.

**Comment by golubenkonick** [comment]
For users running Fedora LXQt with Miriway:
Even with MIRIWAY_SYSTEM_LOCALE1_KEYMAP=1 set in ~/.config/lxqt/session.conf, the variable does not reach the miriway-shell process because the session wrapper /usr/share/lxqt/wayland/miriway/lxqt-miriway-wrapper launches miriway-shell before lxqt-session reads session.conf.
Workaround: manually add the following line to /usr/share/lxqt/wayland/miriway/lxqt-miriway-wrapper before the miriway-shell launch line:

```sh# Enable locale1 keymap tracking (see canonical/mir#3267)
export MIRIWAY_SYSTEM_LOCALE1_KEYMAP=1
```

Proper fix: the lxqt-miriway-session package in Fedora should include this export in the wrapper by default. A bug report has been filed at bugzilla.redhat.com against lxqt-miriway-session.
